### PR TITLE
Alternative approach to LUCENE-8962

### PR DESCRIPTION
**Pull request by s1monw**
this is just a prototype - only intended for discussion

### CODE REVIEW

**Comment by s1monw** [comment]
see #1552 for reference

File: lucene/core/src/java/org/apache/lucene/index/MergePolicy.java [hunk 1]
```diff
@@ -399,8 +423,19 @@ public String segString(Directory dir) {
       }
       return b.toString();
     }
+
+    boolean await(long timeout, TimeUnit unit) {
+      for (OneMerge merge : merges) {
+        if (merge.await(timeout, unit) == false) {
```
Comment (dsmiley) [inline lucene/core/src/java/org/apache/lucene/index/MergePolicy.java:429]: This looks suspicious when there is more than one merge.  Shouldn't the timeout decrease as time is used up by earlier merges?

In practice, when is there more than one?  I've been confused on this matter when I developed a custom MP/MS.

Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/index/MergePolicy.java:429]: in a real change that's correct. in a prototype as this is it's really just there to visualize the idea. I didn't do this on purpose to not discuss impl details. that's not the point of this it's really just a PR to make commenting simpler.

**Comment by msfroh** [comment]
This approach makes sense to me. 

I like how much simpler the addition of MergeSpecification.await() makes things, versus the CountDownLatch hackery of the previous approach. Also, updatePendingMerges returning the MergeSpecification is much cleaner than explicitly computing a merge from within prepareCommitInternal.

**Comment by s1monw** [comment]
@msfroh I opened https://github.com/apache/lucene-solr/pull/1585 to make it easier to do this.

File: lucene/core/src/java/org/apache/lucene/index/MergePolicy.java (hunk 1 above)
Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/index/MergePolicy.java:429]: I fixed this in the followup PR https://github.com/apache/lucene-solr/pull/1585

**Comment by s1monw** [comment]
superseded by #1585

File: lucene/core/src/java/org/apache/lucene/index/IndexWriter.java [hunk 2]
```diff
@@ -2144,39 +2145,43 @@ private synchronized boolean updatePendingMerges(MergePolicy mergePolicy, MergeT
...
         final int numMerges = spec.merges.size();
         for(int i=0;i<numMerges;i++) {
           final MergePolicy.OneMerge merge = spec.merges.get(i);
           merge.maxNumSegments = maxNumSegments;
         }
       }
     } else {
-      spec = mergePolicy.findMerges(trigger, segmentInfos, this);
+      switch (trigger) {
+        case COMMIT:
```
Comment (dsmiley) [inline lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:2170]: There is an inconsistency here that suggests something is wrong, or at least confusing enough to deserve a comment.  For case COMMIT, we call a findFullFlushMerges.  Shouldn't it be on case FULL_FLUSH to be consistent with the method we are calling?  Or should findFullFlushMerges be called findCommitMerges?

Comment (mikemccand) [reply lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:2170]: Full flush happens for refresh and commit.

But we have only implemented "merge on commit" so far.

I would love to also add "merge on refresh", but until we do so, I think it's correct for `IndexWriter` to separate out the `COMMIT` case here like this.

**Pull request closed without merging**